# Homework
Set up the environment
You need to install Python, NumPy, Pandas, Matplotlib and Seaborn. For that, you can use the instructions from 06-environment.md.

### Q1. Pandas version

What's the version of Pandas that you installed?

In [2]:
import pandas as pd
import numpy as np

In [3]:
pd.__version__

'2.3.1'

**Getting the data**

For this homework, we'll use the Car Fuel Efficiency dataset. Download it from here.

You can do it with wget:

wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv
Or just open it with your browser and click "Save as...".

Now read it with Pandas.

In [4]:
df = pd.read_csv("car_fuel_efficiency.csv") ## data was saved from browser as .csv and uploaded into jupyter.

In [5]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

### Q2. Records count
How many records are in the dataset?

In [6]:
df.count()

engine_displacement    9704
num_cylinders          9222
horsepower             8996
vehicle_weight         9704
acceleration           8774
model_year             9704
origin                 9704
fuel_type              9704
drivetrain             9704
num_doors              9202
fuel_efficiency_mpg    9704
dtype: int64

In [7]:
df.describe().round(2)

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
count              9704.00        9222.00     8996.00         9704.00   
mean                199.71           3.96      149.66         3001.28   
std                  49.46           2.00       29.88          497.89   
min                  10.00           0.00       37.00          952.68   
25%                 170.00           3.00      130.00         2666.25   
50%                 200.00           4.00      149.00         2993.23   
75%                 230.00           5.00      170.00         3334.96   
max                 380.00          13.00      271.00         4739.08   

       acceleration  model_year  num_doors  fuel_efficiency_mpg  
count       8774.00     9704.00    9202.00              9704.00  
mean          15.02     2011.48      -0.01                14.99  
std            2.51        6.66       1.05                 2.56  
min            6.00     2000.00      -4.00                 6.20  
25%           13.30     2006.00      -1.00                13.27  
50%           15.00     2012.00       0.00                15.01  
75%           16.70     2017.00       1.00                16.71  
max           24.30     2023.00       4.00                25.97

### Q3. Fuel types
How many fuel types are presented in the dataset?

In [8]:
df["fuel_type"].unique()

array(['Gasoline', 'Diesel'], dtype=object)

In [9]:
df["fuel_type"].nunique()

2

### Q4. Missing values
How many columns in the dataset have missing values?

In [10]:
df.isnull().sum()

## from the results we see that there are 4 columns with missing values.

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

### Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [11]:
max_fuel_eff_Asia = df[df["origin"] == "Asia"]["fuel_efficiency_mpg"].max()
max_fuel_eff_Asia

np.float64(23.759122836520497)

In [12]:
max_fuel_eff_by_origin = df.groupby("origin")["fuel_efficiency_mpg"].max()
# Then access Asia specifically:
max_fuel_eff_Asia_alt = max_fuel_eff_by_origin["Asia"]
max_fuel_eff_Asia_alt

np.float64(23.759122836520497)

### Q6. Median value of horsepower
Find the median value of horsepower column in the dataset.
Next, calculate the most frequent value of the same horsepower column.
Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
Now, calculate the median value of horsepower once again.
Has it changed?

- Yes, it increased
- Yes, it decreased
- No


In [13]:
median_horsepower = df["horsepower"].median()
mode_horsepower = df["horsepower"].mode()[0] ## the first index, ie [0], is added to the end of the mode function because the result returns a series.
median_horsepower, mode_horsepower

(np.float64(149.0), np.float64(152.0))

In [14]:
df['horsepower'] = df["horsepower"].fillna(mode_horsepower)

In [15]:
median_horsepower2 = df["horsepower"].median()
median_horsepower2

np.float64(152.0)

In [16]:
df["horsepower"].isnull().sum()

np.int64(0)

##### hence we can see that the median horsepower increased, taken the value of 152, ie the mode which has replaced missing/null values

### Q7. Sum of weights
1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result?

In [17]:
# Step 1: Select all cars from Asia
asian_cars = df[df['origin'] == 'Asia']

# Step 2: Select only the specified columns
selected_columns = asian_cars[['vehicle_weight', 'model_year']]

# Step 3: Select the first 7 rows
result = selected_columns.head(7)
result

## alternatively: Chain the operations, gives the same result.
## result = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7) 
## or Asian cars = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7)

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [18]:
## Step 4: Get the underlying NumPy array. Let's call it X. 
asian_cars = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7)
X = asian_cars.to_numpy()
X

## Alternatively,
## X = result.values

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [33]:
## Step 5: Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
#matrix multiplication follows .dot()
B = X.T
XTX = B.dot(X)
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [34]:
##Step 6: Invert XTX
Inv_XTX = np.linalg.inv(XTX)
Inv_XTX

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [35]:
## Step 7:Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [36]:
#step 8; Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
Z = Inv_XTX.dot(B)
Z

array([[-1.31202622e-04, -8.63909858e-05,  3.72634923e-04,
        -4.02726650e-04, -1.62513724e-04, -5.52342829e-05,
         4.65094049e-04],
       [ 2.62636846e-04,  1.96990690e-04, -4.73392228e-04,
         6.58944477e-04,  3.08357831e-04,  1.51636137e-04,
        -6.07979633e-04]])

In [37]:
w = Z.dot(y)
w

array([0.01386421, 0.5049067 ])

In [39]:
#### step 9; sum of all the elements of the result
Sum = w[0] + w[1]
Sum

np.float64(0.5187709081074016)